addons/mail/models/discuss/ir_websocket.py

```diff
@@ -22,10 +24,18 @@ def _get_im_status(self, data):
     @add_guest_to_context
     def _build_bus_channel_list(self, channels):
         channels = list(channels)  # do not alter original list
+        discuss_channel_ids = list()
+        for channel in channels.copy():
```

Comment:
question remains, why do we need a copy here?

Reply:
I can not iterate and remove elements from a list at the same time. Theses are just workarounds, I could also remove undesirable elements after the loop.

```python
channels = [1, 2, 3, 4, 5]

for channel in channels:
    print(channel) # 1, 2, 3, 5
    if channel == 3:
        channels.remove(channel)

for channel in channels.copy():
    print(channel) # 1, 2, 3, 4, 5
    if channel == 3:
        channels.remove(channel)

for channel in reversed(channels):
    print(channel) # 5, 4, 3, 2, 1
    if channel == 3:
        channels.remove(channel)
```

Reply:
Ok I understand now, I guess it's fine to keep it this way!

Reply:
two cents, we have a guideline to prefer `list(channels)` over `channels.copy()`
```suggestion
        for channel in list(channels):
```

Reply:
Suggestion:
```py
from odoo.tools import partition
```
```py
    def _is_discuss_channel(self, channel):
        return isinstance(channel, str) and channel.startswith('discuss.channel_')

    def _build_bus_channel_list(self, channels):
        discuss_channels, channels = partition(self._is_discuss_channel, channels)
        discuss_channel_ids = [
            int(channel.removeprefix('discuss.channel_'))
            for channel in discuss_channels
        ]
        guest = self.env["mail.guest"]._get_guest_from_context()
        if guest:
            channels.append(guest)
        domain = ["|", ("is_member", "=", True), ("id", "in", discuss_channel_ids)]
        channels.extend(self.env["discuss.channel"].search(domain))
        return super()._build_bus_channel_list(channels)
```